src/clj/commiteth/eth/core.clj

```diff
@@ -51,59 +53,94 @@
                           {:password password :file-path file-path}))))))
 
-(defn get-web3j-nonce [web3j-instance]
-  (.. (.ethGetTransactionCount web3j-instance (env :eth-account) DefaultBlockParameterName/LATEST)
+(defn get-nonce []
+  (.. (.ethGetTransactionCount @web3j-obj 
+                               (env :eth-account) 
+                               DefaultBlockParameterName/PENDING)
       sendAsync
       get
       getTransactionCount))
 
-(defprotocol INonceTracker
-  "The reason we need this is that we might send mutliple identical
-  transactions (e.g. bounty contract deploy with same owner) shortly
-  after another. In this case web3j's TX counting only increases once
-  a transaction has been confirmed and so we send multiple identical
-  transactions with the same nonce.
-
-  Web3j also provides a TransactionManager that increases nonces but it
-  does not track nonces related to transactions and so as far as I understand
-  it might cause transactions to be executed twice if they are retried.
-
-  https://github.com/web3j/web3j/blob/d19855475aa6620a7e93523bd9ede26ca50ed042/core/src/main/java/org/web3j/tx/RawTransactionManager.java"
-  (get-nonce [this internal-tx-id]
-    "Return the to be used nonce for an OpenBounty Ethereum
-    transaction identified by `internal-tx-id`. As these IDs are stable
-    we can use them to use consistent nonces for the same transaction."))
-
-(defrecord NonceTracker [state]
-  INonceTracker
-  (get-nonce [this internal-tx-id]
-    (let [prev-nonce  (get @state internal-tx-id)
-          web3j-nonce (get-web3j-nonce @web3j-obj)
-          nonces      (set (vals @state))
-          nonce       (if (seq nonces)
-                        (inc (apply max nonces))
-                        web3j-nonce)]
-      (when prev-nonce
-        (log/warnf "%s: tx will be retried (prev-nonce: %s, new-nonce: %s, web3j-nonce: %s)"
-                   internal-tx-id prev-nonce nonce web3j-nonce))
-      ;; TODO this is a memory leak since tracking state is never pruned
-      ;; Since we're not doing 1000s of transactions every day yet we can
-      ;; probably defer worrying about this until a bit later
-      (swap! state assoc internal-tx-id nonce)
-      nonce)))
-
-(def nonce-tracker
-  (->NonceTracker (atom {})))
-
-(defn get-signed-tx [{:keys [gas-price gas-limit to data internal-tx-id]}]
-  "Create a sign a raw transaction. 'From' argument is not needed as it's already encoded in credentials.
-   See https://web3j.readthedocs.io/en/latest/transactions.html#offline-transaction-signing"
-  (let [nonce (get-nonce nonce-tracker internal-tx-id)]
-    (log/infof "%s: Signing nonce: %s, gas-price: %s, gas-limit: %s"
-               internal-tx-id nonce gas-price gas-limit)
-    (-> (RawTransaction/createTransaction (biginteger nonce) gas-price gas-limit to data)
-        (TransactionEncoder/signMessage (creds))
-        (Numeric/toHexString))))
+
+(defprotocol TxTracker
```

Comment:
I like to prefix protocols with `I` so it's more apparent when they appear in stack traces. Just for your consideration :)

Reply:
Renamed.